src/main/java/parking/services/PLNPricing.java

```diff
@@ -0,0 +1,35 @@
+package parking.services;
+
+import org.springframework.stereotype.Service;
+import parking.entities.Plate;
+import parking.entities.enums.Region;
+
+import java.math.BigDecimal;
+import java.time.Duration;
+import java.util.Objects;
+
+@Service
+public class PLNPricing extends AbstractRegionPricing {
+
+    private static final BigDecimal FIRST_HOUR_VIP = BigDecimal.valueOf(0);
+    private static final BigDecimal SECOND_HOUR_VIP = BigDecimal.valueOf(2);
+    private static final BigDecimal NEXT_HOUR_VIP = BigDecimal.valueOf(1.2);
+
+    private static final BigDecimal FIRST_HOUR = BigDecimal.valueOf(1);
+    private static final BigDecimal SECOND_HOUR = BigDecimal.valueOf(2);
+    private static final BigDecimal NEXT_HOUR = BigDecimal.valueOf(1.5);
+
+    @Override
+    public boolean matchRegion(Plate plate) {
+        return Objects.equals(Region.PLN, plate.getRegion());
+    }
+
+    @Override
+    public BigDecimal calculatePrice(Plate plate) {
+        long duration = Duration.between(plate.getStart(), plate.getEnd()).toHours();
+
+        if (plate.isVip())
```

Comment:
1. Czy tutaj można by zastosować wzorzec strategii? Czy chciałbyś tak zrobić?
2. Dla 2h powinniśmy zwrócić 3zł czy 2zł?

Reply:
1. Tam w zasadzie już jest zastosowany wzorzec strategii (PricingService i zastosowanie metody calculatePrice(Plate)) i PLNPricing jest konkretną strategią, gdzie AbstractRegionPricing jest "interfejsem" (tutaj klasą abstrakcyjną) strategii.
Myślę tylko, że gdybym zaimplementował np DollarPricing, czy EuroPricing, duplikowałby się kod metody calculatePrice, więc mógłbym ją wyodrębnić i wcisnąć gdzieś np. do dziedziczonej klasy abstrakcyjnej. Wtedy konkretne strategie ograniczałyby się jedynie do przekazania swoich wartości.
Mógłbym wtedy też, zamienić klasę abstrakcyjną na singleton, który za jedno z pól przetrzymuje strategię obliczania opłaty. Klasy dziedziczące po tej klasie (konkretne strategie po tym singletonie "strategii",  które reprezentują walutę) w swoim rozwiązaniu podawałyby swoje przeliczniki.
W ten sposób uruchamiając aplikację musiałbym jedynie pamiętać o tym, aby zainicjalizować pole singletona "strategii" jakąś konkretną strategią przeliczania opłaty.

2. 
![image](https://user-images.githubusercontent.com/6865620/42352032-63aa683e-80b8-11e8-9b12-d049d23313e2.png)
Hmmm, teraz jak na to patrzę to wychodziłoby, że za pierwszą godzinę płaci się 1 PLN, za drugą 2 PLN, co znaczy, że za 2 godziny płaci się łącznie 3 PLN. Pierwotnie pomyślałem, że 1 godzina = 1 PLN, 2 h = 2 PLN, etc. Pomyliłem się.